rfcs/text/1020-link-entities.md

```diff
@@ -0,0 +1,385 @@
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+## Extending the Entity Type definition to allow for Link Entity Types
+
+### The marker annotation
+
+It will be necessary to define some form of indicator inside an entity type schema to identify it as a link entity type.
+Keeping in mind the potential of future type inheritance, this RFC opts to achieve this by allowing an additional field on entity types which conforms to the following JSON schema:
+
+```json5
+"allOf": {
+  "type": "array",
+  "items": {
+    "type": "object",
+    "properties": {
+      "$ref": {
+        "type": "string",
+        "const": "https://blockprotocol.org/@blockprotocol/types/entity-type/link/v/1"
+      }
+    },
+    "required": ["$ref"]
+    "additionalProperties": false
+  }
+},
+```
+
+At the present, the only reference allowed within the `allOf` will be a reference to an entity type with the following schema:
+
+```json5
+{
+  kind: "entityType",
+  $id: "https://blockprotocol.org/@blockprotocol/types/entity-type/link/v/1",
+  type: "object",
+  title: "Link",
+  properties: {},
+  additionalProperties: false,
+}
+```
+
+Seeing the presence of
+
+```json5
+"allOf": [
+  {
+    "ref": "https://blockprotocol.org/@blockprotocol/types/entity-type/link/v/1"
+  }
+]
+```
+
+within an entity type will be an indicator that it is a link entity type.
+
+### An example
+
+In the future, this `allOf` array may be used for inheritance of types, whereby it would be necessary to resolve the types that are inherited from to see if any of those are a link entity type ([see the RFC proposing an inheritance mechanism](https://github.com/blockprotocol/blockprotocol/pull/428)).
+For now, it is a simple marker annotation.
+
+This is an example of a valid link entity type under the proposal:
+
+```json5
+{
+  kind: "entityType",
+  $id: "https://example.com/@example/types/entity-type/marriage/v/1",
+  type: "object",
+  title: "Marriage",
+  allOf: [
+    {
+      ref: "https://blockprotocol.org/@blockprotocol/types/entity-type/link/v/1",
+    },
+  ],
+  properties: {},
+  additionalProperties: false,
+}
+```
+
+## Extending the Entity definition to allow for Link Entity Types
+
+### `linkData`
+
+The definition of an entity is extended to allow for the presence of a `"linkData"` field which conforms to the following schema:
+
+```json5
+"linkData": {
+  "type": "object",
+  "properties": {
+    "leftEntityId": {
+      "type": "string",
+    },
+    "rightEntityId": {
+      "type": "string",
+    },
+    "leftToRightOrder": {
+      "type": "integer",
+      "minimum": 0
+    },
+    "rightToLeftOrder": {
+      "type": "integer",
+      "minimum": 0
+    }
+  },
+  "required": ["leftEntityId", "rightEntityId"],
+  "additionalProperties": false
+}
+```
+
+#### "Left" and "Right"
+
+At the present, the system specifies uni-directional links, that is, links that go in a single direction from one entity to another entity.
```

Comment:
I don't think uni/bi/multidirectionality need hyphenating. I know we decided this for bi/multitemporality. Cc @CiaranMn

Reply:
Google seems to suggest it does not, but I wouldn't worry _toooo_ much about grammar, this is just an RFC 😅 

```suggestion
At the present, the system specifies unidirectional links, that is, links that go in a single direction from one entity to another entity.
```